### PR TITLE
Build wheels in GitHub actions

### PR DESCRIPTION
**Pull request by mtreinish**
This PR adds CI jobs in github actions to build release wheels for linux, macOS, and windows using github actions. The idea behind it is that at release time we can use github actions to trigger wheels that are distributed on pypi with precompiled binares preventing the need for users to compile scs locally for their environments every time they go to install scs. It uses [cibuildwheel](https://github.com/joerick/cibuildwheel) to perform the wheel builds which automates most of the build process so we create binaries for all the environments.

Right now the PR is running these builds on every commit, but can/should be switched to just be triggered by tags/releases. They also can be setup to automatically upload the binaries to pypi with twine if desired (this is typically the configuration I used for releases).

TODO:

Right now the windows builds are not working all the tests return NaN instead of the expected result: https://github.com/mtreinish/scs-python/runs/1524626020?check_suite_focus=true#step:9:73 The lengthy git log for the PR is from all of my attempts to get this working on windows. I tried using cibuildwheel at first but hit this same issue and decided to try switching to use conda and python directly to have more control over the build process. This will need to be fixed so we can also distribute precompiled binaries for windows users.

### CODE REVIEW

**Comment by bodono** [comment]
This is great, thank you for putting this all together!

I like the idea of automatically uploading the binaries, that makes everything very simple. I will think about whether every commit or just releases is better, for now what you have is fine.

For the windows builds, it looks like the SDP problems are the failing ones. For SDPs we need to link against blas and lapack libraries (for the eigendecomposition routines). I notice a few error messages in the logs that suggest it can't find the lapack / blas libs, eg, [here](https://github.com/mtreinish/scs-python/runs/1524626020?check_suite_focus=true#step:7:136). The way SCS tries to link against those libs is by using numpy's `get_info' [method](https://github.com/bodono/scs-python/blob/master/setup.py#L87). It seems that in the case you posted numpy thinks it is linked against MKL, but that library is actually missing. There are a couple of options here, one is to use a different numpy wheel that is hopefully pointing to the right place.  Another is to install MKL (I think there is a conda version) and that might just work out of the box. The other is to manually install lapack / blas and then point SCS to the manually installed ones (via the environment variable used [here](https://github.com/bodono/scs-python/blob/master/setup.py#L57), though I'm not sure how that would work in the distributed wheel.

**Comment by bodono** [comment]
Hi Matthew, do you want me to pick this up (I don't really have cycles now but might soon hopefully), or are you planning on finishing it? I really want to get this checked in as I think it's really useful!

File: tools/openblas_support.py [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+import hashlib
+
+from tempfile import mkstemp, gettempdir
+from urllib.request import urlopen, Request
+from urllib.error import HTTPError
+import zipfile
+import tarfile
+
+OPENBLAS_V = 'v0.3.9'
+OPENBLAS_LONG = 'v0.3.9'
```
Comment (h-vetinari) [inline tools/openblas_support.py:16]: This is a pretty old openblas version; there's 0.3.15 in the meantime, see [here](https://github.com/xianyi/OpenBLAS/releases)

**Comment by bodono** [comment]
Hi @mtreinish , are you still working on getting this merged in?

**Comment by mtreinish** [comment]
> Hi @mtreinish , are you still working on getting this merged in?

Yeah, I was heads down on other work for a while but recently I picked it back and have been trying to fix the windows builds (which are the only thing not working still) but not having much luck.

**Comment by bodono** [comment]
Ok great, thanks for putting in the work. If you ever feel like you want to give up on the windows builds and get this merged anyway, let me know!

**Comment by bodono** [comment]
I finally managed to get windows working with github actions and passing tests. The yaml file is [here](https://github.com/bodono/scs-python/blob/master/.github/workflows/build.yml). Does this help you overcome the windows issues at all?

**Comment by mtreinish** [comment]
@bodono thanks that was enough of a hint I was able to get the windows builds working: https://github.com/mtreinish/scs-python/runs/2766833786?check_suite_focus=true

I think this is ready for review now. There are 2 small things though, first I couldn't get win32 binaries to be built because of https://github.com/conda-incubator/setup-miniconda/issues/166 I couldn't initialize a 32bit conda env. If that ever gets fixed it would be good to add 32bit builds since a lot of windows users end up with a 32bit version of python (even if they're on a win64 OS). The second is right now the wheels are built and just uploaded to the gha artifacts for the workflow so to publish the wheels you'd have to download them from the artifacts and then upload to pypi manually. I personally typically setup the jobs to also publish to pypi. For example: https://github.com/Qiskit/retworkx/blob/main/.github/workflows/wheels.yml#L65-L69 (although that only runs on tags not every commit). I can adjust this however you prefer

**Comment by h-vetinari** [comment]
Nice progress @mtreinish! Might have to merge master or rebase though - appveyor is [gone](https://github.com/bodono/scs-python/commit/35956814d7d55357da3e1316d924b50085af1be7) now.

> If that ever gets fixed it would be good to add 32bit builds since a lot of windows users end up with a 32bit version of python (even if they're on a win64 OS).

I'd be interested where you get that impression (or data!). Confusingly, the API that `sys.platform` shows is `Win32` even on 64bit python-runtimes. 32bit builds have not been built by anaconda for years already, the amount of relevant 32bit-only windows hardware out there is probably less than the python 2 userbase these days (this was an issue until very recently why PyPy couldn't be built for windows in conda). I see that cpython still provides binaries, but couldn't find download stats because pypistats doesn't distinguish the bitness of windows.

**Comment by bodono** [comment]
Sorry I took so long on this, LGTM.

**Pull request merged**